### ISSUE TITLE
Regionally distribute demand and supply

### ISSUE THREAD

**Issue by brynpickering** [calliope-project/euro-calliope#79]
Datasets are rarely available with the same resolution as we need in Euro-Calliope layers. Instead, we distribute e.g. national data to sub-national regions using datasets like population. There is much scope to increase the accuracy of this regionalisation, using various datasets from e.g. Eurostat. But, all these datasets need to be applicable at any layer resolution. E.g. if using arbitrary shapefiles as the layer resolution (#76), it should still be possible to use a NUTS3 resolved dataset to regionalise demand/supply data. 

Accordingly, there needs to be configuration options to link demands to regionalisation datasets (e.g. residential electricity demand is regionalised using population) and all possible regionalisation datasets need to be rasterised so they can be easily masked and applied to create weighted averages.

Datasets this applies to include: population (residential), Eurostat freight loading/unloading (industry subsectors), Gross value added (commercial), freight loading/unloading (industry subsectors), EU ETS data (some industry subsectors). See [here](https://github.com/calliope-project/euro-calliope-2.0/blob/main/src/construct/annual_subnational_demand.py) for its current application in the lost sibling.

**Comment by brynpickering** [comment]
This links to and expands the scope of bringing the relative residential/industrial electricity demand from `solar-and-wind-potentials`, as discussed in #63

**Comment by timtroendle** [comment]
What exactly do we need to do here? What I see is: 

1. Allow for custom shapes,
2. Integrate renewable potentials (handled in #77),
3. Integrate industry demand (handled in #63).

So am I right in saying that this issue is solved when (1) is solved? If not, what else?

**Comment by brynpickering** [comment]
There's a few other things in addition. This includes regionalising service-sector demand (e.g. commercial building heat demand) based on service-sector gross-added value (this *isn't* industry demand) and (possibly) regionalising marine and aviation fuel demand based on activity of airports and shipping ports (I don't do this at present, but use industry demand as a proxy for possible fuel generation hot-spots). Allowing for custom shapes (1) isn't strictly necessary here, but rather that this process should be shape-agnostic, as it is with the current population/EU-ETS method.

**Comment by timtroendle** [comment]
I don't understand what this issue is about. This workflow is shape-agnostic, apart from the three points mentioned above. When we add things to this workflow, those things should of course be shape-agnostic.

**Comment by brynpickering** [comment]
Yes, all things should be added in a shape-agnostic fashion to the workflow. However, they partially aren't in the sector-coupled sibling, hence the initial driver for me to open this issue in preparation for [the upcoming code sprint](https://github.com/calliope-project/euro-calliope/projects/1). But it's also a question of how we expose this regionalisation to the user. One could regionalise everything based on population, or use only Eurostat employment data (and not EU-ETS) to regionalise industry sub-sectors, etc. Regionalisation could be baked in or it could be configurable, which is the second component of this issue. 

I suspect we don't care about the configurability, so I will update this issue to focus only on the need to handle rasterisation and combination of Eurostat data (which comes in at NUTS2-3 resolution) for use in regionalisation.

**Comment by brynpickering** [comment]
@timtroendle does this now make more sense?

**Comment by timtroendle** [comment]
Yes. Is it correct to say that we do not have an explicit "Allow for custom shapes" issue at this point?

**Comment by brynpickering** [comment]
> Yes. Is it correct to say that we do not have an explicit "Allow for custom shapes" issue at this point?

#76

**Comment by sjpfenninger** [comment]
Comment from Bryn's handover document:

> "This can be done using a geopandas overlay between the spatial datasets that are already included and the geographical units of the resolution of interest. The key here is having a way to subnationalise a particular bit of data on the fly. This requires configuration options to map subsectors to distribution methods (e.g. household building heat: population, commercial building heat: gva, industry steel: EU-ETS, industry textiles and leather: freight & employees) so that if you ask for the subnational data on one of them it can do things automatically. However, it would also ideally only do the overlay of one spatial dataset with one model resolution once, since the operation can be costly. So perhaps first there needs to be this overlay and then that is saved to file and used by all future rules that need it."
